src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/CSharpVirtualDocumentPublisher.cs

```diff
@@ -11,25 +11,37 @@
 using Microsoft.VisualStudio.Text;
 using Microsoft.CodeAnalysis.ExternalAccess.Razor;
 using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+using System.Collections.Immutable;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.LanguageServerClient.Razor.HtmlCSharp;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.LanguageServer;
+using System.Diagnostics;
+using System.IO;
 
 namespace Microsoft.VisualStudio.LanguageServerClient.Razor
 {
     [Shared]
     [Export(typeof(LSPDocumentManagerChangeTrigger))]
     internal class CSharpVirtualDocumentPublisher : LSPDocumentManagerChangeTrigger
     {
-        private const string RoslynRazorLanguageServerClientName = "RazorCSharp";
         private readonly RazorDynamicFileInfoProvider _dynamicFileInfoProvider;
+        Lazy<LSPDocumentMappingProvider> _lazyLspDocumentMappingProvider;
 
         [ImportingConstructor]
-        public CSharpVirtualDocumentPublisher(RazorDynamicFileInfoProvider dynamicFileInfoProvider)
+        public CSharpVirtualDocumentPublisher(RazorDynamicFileInfoProvider dynamicFileInfoProvider,
+            Lazy<LSPDocumentMappingProvider> lazyLspDocumentMappingProvider)
```

Comment:
Lazy loading to get around circular dependency. 

Dependency Chain:

`LSPDocumentManager => LSPDocumentMappingProvider => LSPDocumentManagerChangeTrigger  => LSPDocumentManager`

Thanks @ajaybhargavb!

Reply:
Actually I'd make the mapping provider's acquisition of the lsp document manager lazy. It's a much more core-service that many of our pieces will be using and every time we add a change trigger we'll need to do this little dance unless the mapping provider lazily acquires the manager.